### PR TITLE
Add latex support

### PR DESCRIPTION
**Pull request by florianorben**
Adds latex highlighting support via https://www.npmjs.com/package/katex
as requested in
https://mattermost.uservoice.com/forums/306457-general/suggestions/10160757-add-latex-support

Thought this was a pretty cool and unique feature over competitors, so why not :)
Highlighting is triggered by passing `tex` (or `latex`) as the language to a code block

tex

Example:
```
 ````tex
f(x) = \int_{-\infty}^\infty
    \hat f(\xi)\,e^{2 \pi i \xi x}
    \,d\xi
 ````        ---(remove me otherwise breaks highlight)
```

### CODE REVIEW

**Comment by it33** [comment]
Awesome!!

File: web/react/utils/markdown.jsx [hunk 1]
```diff
@@ -69,6 +69,15 @@ class MattermostMarkdownRenderer extends marked.Renderer {
             usedLanguage = 'xml';
         }
 
+        if (usedLanguage && (usedLanguage === 'tex' || usedLanguage === 'tex')) {
```
Comment (jwilander) [inline web/react/utils/markdown.jsx]: I imagine this is a mistake :P `usedLanguage === 'tex' || usedLanguage === 'tex'`

Comment (florianorben) [inline web/react/utils/markdown.jsx]: Oops yes this should have been `usedLanguage === 'tex' || usedLanguage === 'latex'`, fixed :)

**Comment by coreyhulen** [comment]
@it33 this needs license review

katex.min.js

**Comment by florianorben** [comment]
FYI: https://github.com/Khan/KaTeX/blob/master/LICENSE.txt MIT

**Comment by coreyhulen** [comment]
Looks really cool.  Can you add a unit test for similar to the markdown ones?   in textgeneration.go see var FUZZY_STRINGS_POSTS

**Comment by florianorben** [comment]
@coreyhulen Yep will do :)

**Comment by it33** [comment]
License is compatible,

**Comment by crspeller** [comment]
+1 Love it.

**Comment by florianorben** [comment]
@coreyhulen Test added!
Should katex' license be added somewhere?

+This will cause a merge conflict with #1352 since both modify textgeneration.go at the same location.
Just tell me if I should rebase one of both PRs once one is merged

**Comment by akhmerov** [comment]
Thanks a lot @florianorben! One remark: this does disable using actual latex code with code highlighting, right? While acceptable, it seems a bit odd.

**Comment by florianorben** [comment]
Hey @akhmerov!
The second one being should be `latex`, yes. This has already been fixed, was a stupid typo ;)
https://github.com/mattermost/platform/pull/1351/files#diff-679db143ef363d2321bf05339657483cR72

There actually wasn't any code highlighting for latex blocks before, you can still print it as a code block though, if that is want you meant (just add a code block without specifying latex/tex as a language) (or did you mean something different?):

This

    ```
    f(x) = \int_{-\infty}^\infty
        \hat f(\xi)\,e^{2 \pi i \xi x}
        \,d\xi
    ```
will turn into this:
 ```
f(x) = \int_{-\infty}^\infty
    \hat f(\xi)\,e^{2 \pi i \xi x}
    \,d\xi
 ```

**Comment by akhmerov** [comment]
I didn't realize there was no latex highlighting before.

I'm slightly concerned by this PR breaking markdown syntax as opposed to extending it — triple backquotes aren't *just* code blocks anymore. Would it be possible to use the more common latex delimiters like `$ $` + `$$ $$` or `\( \)` + `\[ \]`? Also is it correct that the PR implements no syntax for inline math?

Either way, these are just minor details, I'm super excited to see latex support and looking forward to 1.3.0.

**Comment by florianorben** [comment]
Hey,

breaking/extending is always a bit in the eye of the beholder I guess - things like markdown tables or github's checklists `- [] foo`  aren't pared of the "spec" too, so I'd say it's fine that way :)

Oh I didn't really know that those delimiters were common to latex - I'll have a look if something like this can be implemented (though the markdown parsing is pretty tricky already, I have to make sure nothing's breaking with that, so it'll take some time).
Inline support should be something like this I suppose `$$\pm\sqrt{a^2 + b^2}$$`? Or what would you suggest (Same "problem" as with blocks though, need to make sure nothing's breaking ;))

**Comment by akhmerov** [comment]
In latex `$ $` or `\( \)` is inline math, `$$ $$` or `\[ \]` is display math. The `$`-based notation is more common I think, but maybe more annoying to parse because it has no difference between opening and closing delimiters.

I would recommend against unconventional use of delimiters, such as using `$$ $$` for inline math.

**Comment by florianorben** [comment]
Cool, thanks for the info, I'm not really that firm with latex so that was useful!
I'm gonna have a look at this quite soon :)

**Comment by akhmerov** [comment]
Thanks for that; please ping me if you have any questions or if you're submitting a PR.

**Comment by florianorben** [comment]
Will do!

**Comment by akhmerov** [comment]
OK, just tried it with the nightly build at https://pre-release.mattermost.com/core/channels/justtesting, and I see that:
* `&` gets incorrectly sanitized. It's used to denote alignment in matrices.
* Several arrows like `\updownarrow` transforms into a pictogram, a conflict with smileys?

I used the examples over here for testing: http://www.intmath.com/cg5/katex-mathjax-comparison.php

**Comment by florianorben** [comment]
Hey,

can you show me an example where `&` doesn't work (though it should)?
I tried it with:
```tex
\begin{pmatrix}
a_{11} & a_{12} & a_{13}\\ 
a_{21} & a_{22} & a_{23}\\ 
a_{31} & a_{32} & a_{33}
\end{pmatrix}
```
And that renders just fine, it's only the code that's marked as *These don't work in KaTeX* [here](http://www.intmath.com/cg5/katex-mathjax-comparison.php) that seem to have problems with `&`.
That would a problem in katex though, we'd need to file a bug/missing feature there (or do you know of any fork etc of katex that implements the missing features?)

Ah good catch with the arrows, that's the smiley lib yes, should be a pretty easy fix, I'll put that on my todo list :)

**Comment by akhmerov** [comment]
Here's what I see in the testing room (saw your message there):

pmatrix

Is it browser- or OS-dependent? I see it both with latest chrome and firefox on linux.

**Comment by florianorben** [comment]
Oh, no it isn't I see it that way too - just didn't really see the `amp` in there (just had a quick look at it and it looked like math ;) ) - Just submitted a WIP PR that fixes that particular issue https://github.com/mattermost/platform/pull/1439.

I'll have a look at the other issues soon :)

**Comment by toogy** [comment]
hi @florianorben , is it possible to add inline support with `$ $` and automatically detect all the `$ $` or `$$ $$`?

i don't know if this is already possible and i need to configure it or something

**Comment by frol** [comment]
Can this feature be documented somewhere, e.g. https://docs.mattermost.com/help/messaging/formatting-text.html?

Also, using ````tex` fence for rendering is not what most of the people might expect. Imagine you want to share a LaTeX snippet with syntax highlight; currently, it will get it rendered as a formula. `$$formula$$` is the usual and expected way to insert a formula.

**Comment by tgy** [comment]
I agree with @frol

**Comment by arjamizo** [comment]
@frol but I think 
```
\`\`\`pdflatex
\`\`\`
```
would be fine :)

**Comment by frol** [comment]
I have found an extensive review of available options for LaTeX formulas in Markdown: https://github.com/cben/mathdown/wiki/math-in-markdown. I hope that somebody learns it and makes a final decision.

@arjamizo While it is quite extensible, it requires some boilerplate; imagine you want to insert a simple formula:

```
 ```pdflatex
  \being{document}$$a=1$$\end{document}
 ```,
```

**Comment by jerlich** [comment]
While the feature is great (math rendering), i will agree with most of the posters, that this is very different than the expected syntax highlighting.  This is more like an REPL: run this code and show the output.

<pre>```latex</pre> should syntax highlight the latex not render the math.

<pre>```math</pre> can be used for rendering math.

I would vote for implementing the pandoc rules. I think they are smart and convenient. (math support should be turned on as a preference).

**Comment by tgy** [comment]
@jerlich very well said

**Comment by michaelcadilhac** [comment]
Hi there.  I've hacked a few things to get inline LaTeX.  The changes are quite light, but it relies on a patched version of **marked.js** by user davitv.
https://github.com/michaelcadilhac/platform/tree/inline-latex
 https://github.com/davitv/marked

I've taken care of having that as an option in the console, but I have no experience whatsoever with pull requests and the such; if anyone wants to apply that, feel free to do so.

Cheers;
M.

Example usage:
schqjbal
sc23l2sw

**Comment by tgy** [comment]
That's very cool, thanks Michael :)

**Comment by michaelcadilhac** [comment]
Has the inline portion ever been integrated?

**Comment by tgy** [comment]
@michaelcadilhac It hasn't

**Comment by michaelcadilhac** [comment]
Is there anything I can do to help with integration?  I host a mattermost where researchers use that feature intensively, but I'm always afraid I'll break everything if I were to update the repository, so I stick to a year-old version…  I'd rather use official releases!

**Comment by jasonblais** [comment]
@michaelcadilhac Are your changes here still up to date https://github.com/mattermost/mattermost-server/compare/master...michaelcadilhac:inline-latex

**Comment by jasonblais** [comment]
Let me check with our design team and see what their thoughts are. Will keep you updated @michaelcadilhac 👍

**Comment by michaelcadilhac** [comment]
@jasonblais Thanks!  Note that this fails for replies; I tried tracking where to add the machinery of bringing `enableInlineLaTeX`, but I got lost in the files.  In particular, [PostMessageContainer](https://github.com/mattermost/mattermost-webapp/search?utf8=%E2%9C%93&q=postmessagecontainer&type=) seems used but never defined.  Any help?

**Comment by michaelcadilhac** [comment]
@jasonblais Any update?

**Comment by jasonblais** [comment]
@michaelcadilhac I'm really sorry for not getting back to you earlier.

I had started a conversation on this, but it unfortunately slipped my mind to follow up. It's now queued for a meeting next week.

**Comment by michaelcadilhac** [comment]
@jasonblais Brilliant, thank you very much!

**Comment by jasonblais** [comment]
Hey @michaelcadilhac! Thanks again for your patience.

First, really appreciate your effort for putting this together. I could definitely see a use case for inline latex across different research groups.

There was a suggestion from the team to consider having the Latex inline support as a plugin - plugins would allow you to override the styling based on the inline Latex syntax, and is easy to maintain across Mattermost releases (no need to rebase your fork every release).

It would use a similar framework to this post-intercept plugin https://github.com/mattermost/mattermost-plugin-profanity-filter, and we just created an intro video for a plugin starter guide: https://www.youtube.com/watch?v=Cx2EBkGkz00&feature=youtu.be

Let me know if that's an option to consider.

There's a big focus on plugins starting tomorrow until Monday [through our plugin hackathon](https://www.meetup.com/mattermost/events/253346351/), so it would be easy for you to connect with people to ask questions if you had interest.

**Comment by jasonblais** [comment]
Thanks @michaelcadilhac! Sounds great. If you do have time in October, we'd be happy to work with you on it

**Comment by michaelcadilhac** [comment]
@jasonblais aaand I'm back.  Say, is the profanity plugin supposed to work against the current master?  When activating, it goes to "This plugin is starting" but then reverts back to "This plugin is not enabled." without signaling any error (including in the logs).

**Comment by jasonblais** [comment]
Yay! Good to see you back :) 

It should work on `master`. Are you activating it in System Console > Management?

**Comment by michaelcadilhac** [comment]
Thanks :-)  Yes I'm using the system console.  Probably a quirk on my installation then, I'll investigate!

**Comment by michaelcadilhac** [comment]
1. I can confirm that implementing a plugin is easy, when starting from a simple backbone (I'm not saying I understand what I did, really.)
2. I've implemented the following plugin: https://github.com/michaelcadilhac/mattermost-plugin-inline-latex
What it does is simply parse LaTeX expressions of the form `$STR$` and `$$STR$$`.  The former is changed to `` `latex STR` `` and the latter to:
```
    ```latex
    STR
    ```
```
3. The codeblock is already piped into KaTeX by the webapp; I had to modify the webapp to also translate codespans, taking care of the `displayMode`:
https://github.com/michaelcadilhac/mattermost-webapp/commit/dc793cd5c7d9511e4008832d1c58b7b3f9354c9c

4. Pros of using plugins rather than the previous approach:
- There's no need to edit the server to create a new option in the admin panel
- The decoupling makes it natural to parse the syntax in the plugin, and render it in the webapp; the tweaked version of marked.js that I was using previously is _less_ needed.

5. Cons:
- The text is actually changed, which means that if there's a LaTeX error, the user will have to edit a code they did not write (since $a$ becomes `` `latex a` ``).
- The message first appears in the main chat window, and is _then_ changed.  Not only is that a bit weird, but since using $$STR$$ _augments_ the height of the message, users have to manually scroll to see the whole message.

6. Conclusion: I'm not sure plugins are the way to go for that functionality. 


What do you think?

**Comment by jasonblais** [comment]
@michaelcadilhac Great evaluation. Let us think about this and get back to you. Appreciate your thoughtfulness here 👍

**Comment by sebastianst** [comment]
Any progress here? rocket.chat supports `\( ... \)` out of the box, btw.
see also https://github.com/mattermost/mattermost-server/pull/1439

**Comment by jasonblais** [comment]
I've touched base with our team on their thoughts. Will follow up 👍

**Comment by jwilander** [comment]
This should definitely be a plugin, rather than a built-in feature. It's almost the perfect example of a feature that makes perfect sense as a plugin. We should strive to improve the plugin framework so that it has any features needed to do something like this just as well as if it were a built-in feature.

First of all, I think this plugin needs to be done on the client-side. What I think is needed to make that possible is a way to hook into the text formatter with a plugin. We have a [MessageWillFormatHook](https://developers.mattermost.com/extend/plugins/webapp/reference/#registerMessageWillFormatHook) already that allows a plugin to modify the plaintext of a post before it runs through our text formatter (a glorified Markdown parser). With something like that it would be fairly trivial to write a plugin to do this just as well as if it were built-in.

@hmhealey any thoughts on hooking into the text formatter with a plugin?

**Comment by michaelcadilhac** [comment]
@jwilander Agreed; conceptually this should be a plugin.  My previous versions were using a modified version of markdown.js that allowed custom parsing rules; this may be the way to go.

**Comment by michaelcadilhac** [comment]
@jasonblais Any news?

**Comment by hmhealey** [comment]
I'm not sure if `MessageWillFormatHook` is sufficient on its own for adding this since any hook into that has to output something that can already be understood by the Markdown formatter. It could be used to add custom latex-block syntax (instead of having it use a code block with the language set to "latex"), but I don't think it can be used to implement inline latex.

There's a few places I think we could let plugins hook into to do this, but since I'm still planning on overhauling the web app's text formatting code and switching it to use commonmark.js, so we need to be careful about what else we add so that it won't conflict with that.

There's a couple ways I think we might be able to do this:
1. If we had a Latex library on the server that could generate an image, we could use the `MessageWillFormatHook` to convert `$something$` into an image like `latex`.
2. The markdown library should be able to render HTML, so if we can figure out a way to whitelist the HTML generated by Katex (or some other Latex library), we could use that same hook to pre-process the Latex before handing it off to the Markdown parser. Marked does have a `sanitizer` option that might be relevant here, although we would need an abstract way for plugins to control that even after we switch to commonmark.js

**Comment by michaelcadilhac** [comment]
Has there been any recent changes on plugins that would allow this?

**Comment by jasonblais** [comment]
@hanzei or @lieut-data would you know?

**Comment by lieut-data** [comment]
I'm not aware of anything that adds the required flexibility here yet.

**Comment by michaelcadilhac** [comment]
Any improvement?  I have to move my classes online, and LaTeX support is crucial for a Slack-like application now.

**Comment by jasonblais** [comment]
@michaelcadilhac There is a community-built plugin for Mattermost from a few months ago - I haven't tested the feature set but might be a starting point at least: https://github.com/kosgrz/mattermost-plugin-latex

**Comment by michaelcadilhac** [comment]
@jasonblais As far as I understand, this is based on an extended interface for plugins to register to receive code blocks (and we could extend that to code spans).  I like the idea, but it seems that the PR has been closed and the associated ticket closed as "won't do".

For the moment, I see _no way_ to have inline LaTeX without hacking the codebase (my point 3 above in https://github.com/mattermost/mattermost-server/pull/1351#issuecomment-427691081).

There's some anecdotal evidence that the users would like that feature right now: last month, a user reported failure using the plugin you mentioned on Github, and yesterday I got an email from a department that was interested in incorporating my plugin—which, again, necessitates hacking Mattermost, and they're usually not comfortable with that.

**Comment by jerlich** [comment]
Gitlab formatted markdown supports inline and blocked math using mathjax. Is there a reason that the same method cannot be used in Mattermost?

**Comment by FreVercauteren** [comment]
Hi @jasonblais 

Any progress on supporting inline LaTeX out of the box (or as a plugin if needed)?  I recently moved my whole research team (80 people) from Slack to Mattermost, and the main question (not to call them complaints :-)) I get repeatedly is why there is no inline LaTeX support?  

Getting this fixed would really help out a lot of people I think...

**Comment by michaelcadilhac** [comment]
Hey @jasonblais ; I'm more than willing to rework all of my changes for that.  I can include @kosgrz framework, so that it's more plugin-friendly; that code I completely understand.  I can extend it to codespans, so that ` ```latex x``` ` would be in a display-math and single quotes would be inline.  That's easy and clean.

However, for $...$ to work, I still need to have access to the markdown tokenizer.  How would you suggest plugins register a new tokenizer (markedjs has [such a feature](https://github.com/markedjs/marked/blob/92998df2ccea5c8587fbe73b6464a86972313fc0/docs/USING_PRO.md))? I'm completely lost with the "registry", the "store", the plugin reducer, and all that...

At the moment, I can either butcher mattermost to add that functionality for the people who _need_ it, or prepare some changes that have a chance of finally being included in the code base.  That'd be my third attempt in as many years at offering changes that would bring inline LaTeX to Mattermost.  Some guidance would really be appreciated to avoid having to do a fourth attempt next year…

**Comment by jerlich** [comment]
In gitlab flavored markdown :
This math is inline $\`a^2+b^2=c^2\`$

**Comment by esethna** [comment]
@michaelcadilhac just to chime in here,

1. If you were to submit this to the webapp as a core feature (not as a plugin), can you summarize (or link to it if you have it) what the code changes would be, and what the expected user experience is that you want to add here? I'd like the dev and UX team to review. 

2. If I'm following this thread correctly it sounds like you are hitting a blocker trying to add this as a plugin due to technical constraints with the plugin infrastructure. If we went forward with the plugin approach to this, would you be open to contributing the required plugin infrastructure changes to support this plugin if our team provided guidance on a technical approach?   

Internally we had a discussion on this and there's some uncertainty on which path to move forward with in the short term. Long term we agree that a plugin makes sense for this feature.

Apologies if some of this was already discussed, just trying to get an understanding of where this is at right now to figure out an approach to move forward. 

cc// @aaronrothschild @agarciamontoro

**Comment by jiucenglou** [comment]
Could you suggest if one could use the `mhchem` package within katex in mattermost ?
The `mhchem` package is conveniently useful to write chemical related content.
